hips/hip-00NN.md

```diff
@@ -0,0 +1,114 @@
+
+## Specification
+
+The `Chart.yaml` should support the following format for `dependencies`:
+
+```
+dependencies:
+- name: "<dependency name>"
+  repository: "git[+<subprotocol>]://<hostname>[:<port>][:][/]<path>[#subdirectory=<path-inside-the-repository>]"
+  version: "<commit-ish>"
+```
+where:
+- `<subprotocol>` is a protocol supported by `git clone` (e.g. `ssh`, `http`, `https`, `file`, etc).
+- `<commit-ish>` is an existing reference (SHA hash, tag or branch name) on the repo.
+- `<path-inside-the-repository>` is the folder where the chart we want installed is located. If not specified, it defaults to the root of the repository.
+  - Implementation note: this has potential for path traversal based security bugs - it needs to be validated and prevented.
+
+Note that `git clone` supports having the `username` and `password` in the repository URL. The implementation of this feature should explicitly forbid that to prevent accidental credential leakage. It should throw an error if the URL contains a `username` or `password`.
+
+For example:
+
+```
+dependencies:
+- name: "jenkins"
+  repository: "git+https://github.com/jenkinsci/helm-charts.git#subdirectory=charts/jenkins"
+  version: "main"
+```
+
+### Installation
+
+When Helm is installing a dependency from git, it should:
+
+- create a temporary directory
+- clone the repo at the specified branch/tag into the temp dir
+  - Helm will require a working git installation to invoke (via subprocess) in order for Helm to utilize chart's git dependencies. Helm will throw an error if git is not installed or misconfigured (e.g. credentials are not set up for private repositories).
+  - for performance reasons, a shallow clone of just the latest commit of a specific branch should be performed (i.e. `git clone --depth 1 --branch <commit-ish> --single-branch --no-tags <repo-url> <temp-dir>`) 
+- treat the cloned git repo similar to a `file:///path/to/temp/dir` style requirement; use `chart.LoadDir` to load that directory (which in turn applied the logic for filtering the files through `.helmignore`) and archives it to `charts/`
+- delete the temp dir
+
+### Linting
+
+`helm lint` should print a warning when a chart contains a git-based dependency, primarily because git references are mutable.
+
+## Backwards compatibility  
+
+This is backwards compatible from Helm perspective - the existing formats for `dependencies` are still supported.
+
+Charts that start using the new format will effectively be changing their minimum required Helm version, i.e. they would be introducing breaking changes and should bump their major version.
+
+## Security implications
```

Comment:
git allows symlinks that can point to anywhere. Helm resolves symlinks and happily packages up the file contents. Is that an additional security risk?

Reply:
Helm does support resolving symlinks during `helm package`: https://github.com/helm/helm/issues/8540 - seems that will be done by design, and it should print a warning.

So does `helm template .`. 

However the symlinks are ignored if they exist inside the tarball itself (i.e. if you craft the tarball manually, and include a symlink, then it will not be followed) - I hope that's by design too 😁 

I've switched the code in the implementation to effectively use the same code path as `helm package` (primarily to make sure that if you're installing a chart from git, it would be have as close as possible to it have been `helm package`d and published into a registry). This means that symlinks get followed and resolved during `helm dep up`, but you do get a warning:

```
walk.go:74: found symbolic link in path: /var/folders/j3/8m3y8q4176118dng0hvdqfh00000gp/T/helm-git-3323594338/templates/secret.yaml resolves to /Users/Dominykas_Blyze/devel/experiments/helm-test/templates/secret.yaml. Contents of linked file included and used
```

Given that `helm package`'s warning was deemed to be enough - I figure this is also enough? I will add an explicit note under "security implications" to explain this - and possibly that should be left at that?